docs/modules/extensions/pages/processors/tree-processor.adoc

```diff
@@ -1,5 +1,38 @@
-= Tree Processor Extension Example
-:navtitle: Tree Processor
+= Tree Processor
+
+TreeProcessor is a DocumentProcessor
+
+== DSL methods
+
+include::page$processor.adoc[tag=docproc]
+
+== Process method
+
+`def process document`
+
+The process method is called from a Document after parsing.
+Therefore document attributes are available and the block structure of the document is available.
+
+The `process` method may modify the document and return the document or nil, or return a new Document.
+The document attributes are reset to the header attributes before the first tree processor is called.
+A tree processor should thus assure that the attributes of the original or returned document are reset to the header attributes by calling `document.restore_attributes` if the attributes have been modified.
+
+There are two basic strategies for a TreeProcessor.
+
+=== Query the document blocks using the `document.find_by` method
+
+This provides quick simple access to the blocks of interest, but does not provide access to the attributes set inline in the document.
+Only the header attributes are available.
+
+A block found this way can be replaced with `parent.blocks[parent.blocks.index original_block] = new_block`.
+
+=== Traverse the block tree
+
+The code here is more complicated and needs to be recursive to find all blocks, but the attributes present at each block may be computed as the traversal proceeds.
+To replay the attributes, for each block execute `document.playback_attributes block.attributes`.
+Remember to call `document.restore_attributes` before returning.
```

Comment:
I think it might be good to have an example of this -- is this so you can determine the state of an attribute at an arbitrary position in the document?

Reply:
I think you're right about needing an example.

Reply:
The subject of tree traversal deserves its own page. It's an important topic, but also an advanced one. That page can set up the scenario when tree traversal is needed and go over the caveats of doing it. It's not something that is really part of the standard Asciidoctor usage and you need to know that you are offroading it a bit (at least as it stands now).

Reply:
I've been looking for examples of when tree traversal is needed and the only one I can come up with is fixing STEM processing, which ought to be doable with a block processor rather than a tree processor anyway. Deemphasizing tree traversal is definitely a good idea.